File: Algorithm/reverse_linked_list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution:
    def reverseList1(self, head):

        if  (not head) or (not head.next):
            return head
        q = self.reverseList1(head.next)
        head.next.next = head
        head.next = None
        return q
    def reverseList2(self, head):
        pre = None
        cur = head
        while cur:
            p = cur.next # 把当前的指针链接到前一个元素
            cur.next = pre
            pre = cur
            cur = p
        return pre
    def reverseList3(self, head):
        
        def reverse(head,newHead):
            if not head:
                return newHead
            cur = head.next
            head.next = newHead
            return reverse(cur,head)
            
        return reverse(head,None)
        
    def reverseList(self, head):
        dummy = ListNode(-1)
        cur = None
        dummy.next = cur
        while head :
            q = head.next # 防止断链
            dummy.next = head
            head.next = cur
            cur = head
            head = q
        return dummy.next
 
                


if __name__ == '__main__':
    sol = Solution()
    lst = [1,2,6,3,5,4]#
    #lst = [1,2]
    head=p= ListNode(-1)
    for value in lst:
        p.next = ListNode(value)
        p = p.next
    p.next = None
    p = sol.reverseList2(head.next)
    print 'fuck'
    while p :
        print p.val
        p = p.next
    print 'das'

File: Algorithm/CPP/reverse_linked_list.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
using namespace std;
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class solution {
public:
	ListNode* reverseList1(ListNode* head) {
		if (head == NULL || head->next == NULL)
			return head;
		auto q = reverseList1(head->next); // ·ÀÖ¹¶ÏÁ´
		head->next->next = head;
		head->next = NULL;
		return q;
	}
	ListNode* reverseList(ListNode* head){
		if (head == NULL || head->next == NULL)
			return head;
		auto dummy = new ListNode(-1);
		auto q = dummy;
		q->next = NULL;
		while (head)
		{
			auto tmp = head->next;
			head->next = q->next;
			q->next = head;
			head = tmp;
		}
		return dummy->next;
	}
};


int main()
{
	ListNode *head = new ListNode(-1);
	auto p = head;
	vector<int> vec{ 1, 2, 3, 5, 9, 8 };
	for (auto x : vec)
	{
		p->next = new ListNode(x);
		p = p->next;
	}
	p->next = NULL;
	//p = root->next;
	//while (p!=NULL)
	//{
	//	cout << p->val << endl;
	//	p = p->next;
	//}
	solution sol;
	auto q = sol.reverseList(head);
	while (q)
	{
		cout << q->val;
		q = q->next;
	}
	//cout << sol.reverseList(head);
	cin.get();
	return 0;
}
